### DISCUSSION TITLE
How can we get the requirement for local installation of `remark` removed?

### DISCUSSION THREAD

**Discussion by halostatue** [remarkjs/remark#960] [Q&A] [answered]
I have a _lot_ of Markdown files where I don’t use Node. The requirement of having to have `remark` installed in order to use `remark-language-server` is…unhelpful.

I have seen some issues and PRs for `remark-language-server` that _may_ resolve this, but as of v 1.1.0, this is still a problem.

```
Cannot turn on language server without `remark` locally. Run `npm install remark` to enable it
```

I’m using `remark-language-server` through `vim-lsp-settings`.

(Frankly, I consider this a bug, but the “bug report” and "feature request" forms on `remark-language-server` appear to be designed to reduce bugs reported.)

**Comment by wooorm** [comment]
We use forms because people don’t provide needed info. And this is an example: we can help you much better if you provide more info.

I personally agree with you that some folks want to use `remark-language-server` without installing `remark` locally. Can you make your case for how you are using and what you want there, and in the underlying project? E.g., here: https://github.com/unifiedjs/unified-language-server/pull/33#discussion_r798363006, and https://github.com/unifiedjs/unified-language-server/issues/30.

**Reply by halostatue** [reply]
There really isn’t any more info to give, and that’s why the forms are ultimately user-hostile.

- I am a consumer of `remark-language-server` through `vim-lsp-settings`.
- I use Markdown _everywhere_, not just in projects. It is my main documentation format.
- Out of dozens of projects that I manage or interact with, there are ~5 that are node.js. One of those is with `pnpm`, two are with `yarn`, and two are with `npm`.
- Having to have `remark` installed locally _does not work_ in cases where I am working with Go, Elixir, Ruby, Rust, or just on standalone markdown files. None of these have node trappings and both adding an otherwise-useless `package.json` and modifying the `.gitignore` to ignore this dross is an annoying waste of my time. None of the projects where this message has annoyed me in the last several weeks would even _remotely_ appreciate me adding this code-litter. Nor am I going to centralize all of my documents into a single place with an equally useless node trapping for something that isn’t based on node.
- I expect a language server to work 100% of the time without local modifications to a project, or require the creation of a project. More to the point, I expect a language server to be installed as part of or by the language server package for the editors that I use (which means I may end up with multiple versions of the language server if I use multiple editors; I have the disk space).
- It is _nice_, but not _necessary_ for the language server to use locally specified versions. It is _necessary_ that the language server not require me to modify or create a project when one does not exist or is not necessary.

I’m really not trying to be difficult, but I suspect that the number of people who have `remark` installed locally to a project is quickly going to be dwarfed by the number of people who just want the nicety of a markdown language server without having to install `remark` anywhere. I don’t _like_ node, would rather not _use_ node, and quite honestly I’m only putting anything in this discussion because seeing that message annoys me every single time I see it and I finally got annoyed enough to say something.

**Reply by wooorm** [reply]
You‘re wrong in that you think that `vim-lsp-settings` and `remark-language-server` are a perfect overlapping venn-diagram.
I am surprised by the naïveté in that, because the warning is about `remark`, the fix has to be there, and the conclusion that `vim-lsp-settings` is perfect. In my opinion `vim-lsp-settings` is partially where the fix could be. Because the venn-diagrams are not perfect overlapping venn-diagrams. Some users of `remark-language-server` need other things than what you need.
Otherwise, I agree with your points and I am glad to hear that users voice how they use tools. I’m surprised by your snarky-ness though.

**Reply by halostatue** [reply]
As I said, I’m snarky about this because I finally got excessively annoyed at this warning and then found the bug reporting process to be user hostile. I’m not intending to be unpleasant, but I _do_ recognize that I am coming across that way because I’m annoyed. For that, I’m sorry; as a maintainer, you don’t need hostile users.

I disagree with you on whether the fix is in `vim-lsp-settings`—if `remark-language-server` is a general markdown LS. As far as I can tell, `vim-lsp-settings` expects that a language server is a complete package and just works out of the box‡. It appears that @remcohaszing submitted mattn/vim-lsp-settings#516 and neovim/nvim-lspconfig#1606 and both were merged around the same time. There’s a secondary PR for nvim-lspconfig which will prevent the use of `remark-language-server` if remark isn’t configured, but I think that’s a net _loss_ if the `remark-language-server` is _good_ for linting markdown files.

If `remark-language-server` is not intended as a general markdown LS, then it is worth a PR to either _remove_ it from `vim-lsp-settings` et al. or make it so that it can _only_ work in the context of someone already invested in the `remark` ecosystem (e.g., neovim/nvim-lspconfig#1641). Or it could be reworked so that someone can opt _into_ the `remark` ecosystem without having to do anything more than `npm install -g remark <other-requirements>` and everything just works (including the `remark-language-server`) just as `prettier` does.

‡ This is less true of something like `neoformat` which can require that a _global_ binary be installed, like `prettier` or `prettierd`.

I think there needs to be a good markdown LS. Maybe `remark-language-server` is it. Maybe it isn’t.

**Reply by wooorm** [reply]
> or make it so that it can only work in the context of someone already invested in the remark ecosystem

That’s what I’m getting at with “it’s unclear where this problem is and can be solved, it could be in `vim-lsp-settings`”. If what you’re saying is true for all `vim-lsp-settings` users, than they can make such a decision there.

> Or it could be reworked so that someone can opt into the remark ecosystem without having to do anything more than npm install -g remark <other-requirements> and everything just works (including the remark-language-server) just as prettier does.

This is what the [***info message***](https://github.com/unifiedjs/unified-language-server/blob/dba7db68db8186a969f7affc5bf5a2c4bf7793bc/lib/index.js#L234-L240) you’re seeing is. It could also be *accidental* that `remark` is missing, which is why there’s an informational message: “nothing is happening now but if you want things to happen make sure remark is installed”

> I think there needs to be a good markdown LS.

I agree with @Murderlon that “language server” can be interpreted in different ways. Interestingly, you mostly talk about linting, which I see as but one part of language servers.
I can see formatting being done without remark being installed/configured locally in certain cases.
I’m not so sure about linting w/o local install or configuration.

**Reply by halostatue** [reply]
Then I think that I will be recommending the _removal_ of `remark-language-server` from `vim-lsp-settings` as specified for `markdown`, because it obviously _isn’t_ a general case markdown language server.

Yes, formatting is one of _several_ features offered by language servers, but formatting is generally considered less important than linking and debugging by most estimations.

**Reply by Murderlon** [reply]
Have you read my comment below? It only "isn't" a general case markdown language server in your definition, and `vim-lsp-settings` doesn't necessarily enforce that definition either. For instance, look at eslint, which is also in `vim-lsp-settings`. It doesn't work without a local version and settings.

I'm also uncertain what unopinionated "debugging" would mean in markdown.

**Reply by halostatue** [reply]
I did read your comment below. I disagree with your position, finding it quite unreasonable. What I have seen here is that `remark-language-server` is a `remark` language server (it is in the name) and that you’re uninterested in supporting the case where people don’t want to have to opt into a whole ecosystem of tooling and configuration _on projects and files where they don’t even apply_.

It’s been a while since I’ve done anything with it—and I haven’t done anything with it in terms of `eslint-language-server`—but I seem to recall that `eslint` has _reasonable_ behaviour if (1) it’s installed globally and (2) there are no configuration files. That means that I can edit `random.js` and get _something_ useful out of it. If `eslint-language-server` only works if there’s a local installation and configuration, then I consider that a bug with `eslint-language-server`, because `eslint` does not require those things. It may work _better_ for a given project with those things, but they aren‘t required.

`remark-language-server` never should have been added to `vim-lsp-settings` with the way that it works as a markdown language server, because it gives the impression that it will do something useful, and then spits out a useless, annoying message and apparently won’t do anything half as useful as `markdownlint` through other mechanisms without opting into a tonne of configuration and tooling.

**Reply by Murderlon** [reply]
> I seem to recall that `eslint` has _reasonable_ behaviour if (1) it’s installed globally and (2) there are no configuration files. That means that I can edit `random.js` and get _something_ useful out of it.

What is something useful in this case? AFAIK you won't get anything out of eslint if you don't tell it what to do. Regarding global installs, here is their [take](https://eslint.org/docs/user-guide/getting-started):

> It is also possible to install ESLint globally rather than locally (using npm install eslint --global). However, this is not recommended, and any plugins or shareable configs that you use must be installed locally in either case.

This is the exact same situation with remark. When linting is highly contextualized, with likely big differences between projects, and only opinionated rules — there is no one-size-fits-all. Due note that you are taking a strong stance in this discussion on principle of it not having it work out-of-the-box, but not describing _what_ it would do out of the box. The latter is the problem, because who will agree to what is an error if there are no true errors in markdown?

That doesn't mean nothing can be done out-of-the-box, an interesting example of what it could do is something I coined here: https://github.com/remarkjs/remark/discussions/940

Lastly, whether it belongs in `vim-lsp-settings` is a discussion to be had there.

**Comment by Murderlon** [comment, marked as answer]
I think the definition and workings of a language server can be more generic than batteries-included majority of language servers. This is because most of them, which we have grown accustomed to, take the (opinionated) convention over configuration route. And that makes sense, because LSPs like gopls, tsserver, rust-analyser, etc tell you what's right or wrong, like an unused import. You can take that route because you can assume no one will object to getting an warning about that unused import.

Markdown, however, doesn't really have errors or warnings that everyone agrees on for that matter. So when it comes to having remark-language-server work out-of-the-box, you have to ask yourself what would it lint? or format? There are hundreds of plugins doing all kinds of highly opinionated things.

Secondly, most language servers are used standalone. Meaning its config is only used for that. But with remark it's likely the project at hand already has a config file and is being used by remark node or remark CLI, but you just want to have additional editor hints. It's good to share those settings.

I'm not completely against making remark-language-server work out-of-the-box but I'm afraid it wouldn't do much by default.

**Comment by halostatue** [comment]
This question is fundamentally unanswerable given the discussion above.

**Reply by wooorm** [reply]
I see several potential answers, several of these already mentioned:

- Remove `remark-language-server` from `vim-lsp-settings`
- Open an issue or leave a comment with `vim-lsp-settings` about your frustration with how it works
- Add functionality to `vim-lsp-settings` project to *only* enable `remark-language-server` when there is configuration found
- Ignore this particular information message from `vim-lsp-settings`, or to the same effect: add functionality to `vim-lsp-settings` project to *only* enable `remark-language-server` when `remark` is installed
- Not the most sensical, but `vim-lsp-settings` could package `remark`
- Add an option to `unified-language-server` (which would bubble through `remark-language-server`), to not emit this informational message
- Probably some more? Is this enough?

Finally, your insights in what you believe “remark” is. Or what “remark-language-server” should do. Or whether a hypothetical “markdown-language-server” could/should exist and what *that* would do. These are all welcome and interesting.

---

It feels to me like you’re here to rant and not at all interested in searching for consensus on a solution. You‘re not talking to customer service that owes you something. This is open source.

**Reply by halostatue** [reply]
I do not believe that there _is_ consensus to be reached, and it’s based on the responses that I have been given about what `remark` is _based on comments by you, the `remark` maintainers_. As I said, the initial error appears to have been one of your contributors submitting a `remark-language-server` PR to `vim-lsp-settings` &c. without sufficient guards to prevent it from being used outside of a `remark` project, which is apparently a thing (I had never _heard_ of `remark` prior to this). The second error was mine: I assume the `remark-language-server` was, in fact, a Markdown language server when it clearly _is not_.

It appears that `eslint` _has_ changed from what it used to be (which IMO makes it _less_ useful) in that it used to assume a certain amount of default configuration if none was provided. On the other hand, there are many _extremely_ useful tools which assume defaults (which may or may not be likeable or common) out of the box. `rubocop`, for example, does not require a `.rubocop.yml` in a project to be "useful" (my opinions on Rubocop defaults notwithstanding) and can be used on standalone Ruby files. `markdownlint-cli` or `markdownlint-cli2` can be used on standalone Markdown files. Yet both can _also_ use local configuration files on execution.

A `markdown-language-server` could, in fact, do _many_ useful things for authors. In addition to providing the main functionality of live syntax checking and perhaps provide some navigation assistance for headings and local references, it could potentially go as far as `proselint` in an asynchronous manner. Language servers provide utility to _many_ editors rather than requiring a new plug-in be created for each possible editor because of the work that MS did on the LSP.

I really do wish you the best of luck with whatever it is that `remark` is, but what _I_ want first from any language server is a no-configuration-required asynchronous linter ("diagnostics" in LSP terminology) for the language it advertises to support (only later do I want to consider formatting, &c.). `remark-language-server`, through `vim-lsp-settings` and the `nvim` equivalent, advertised Markdown, and it does not deliver that, but a noisy annoying error message. There’s a fundamental mutual disconnect here between what it _appears_ that you offer and what it _appears_ that I want, and we have been speaking at cross purposes for two days.

**Reply by wooorm** [reply]
> There’s a fundamental mutual disconnect here between what it appears that you offer and what it appears that I want, and we have been speaking at cross purposes for two days.

Right. I feel this disconnect was discovered 2 days ago (Murderlon’s comment, mine around that time too).

> but what I want first from any language server is a no-configuration-required asynchronous linter

If you want those things, I recommend you pick either of my first 4 proposed solutions in `vim-lsp-settings`. You can also work on the other proposed solutions in `remark-language-server`

**Reply by halostatue** [reply]
I have already filed an issue to have `remark-language-server` removed from `vim-lsp-settings`, since it doesn’t do anything useful in the main case. The answer I have put this response on _is_ the correct answer, despite the fact that you guys keep marking it as "not the answer". You’re not interested in providing a general-purpose markdown language server. That’s great. But it never should have been added to `vim-lsp-settings` as a markdown language server in the first place by one of your contributors.

If you guys had been less obtuse and instead said from the beginning "we aren’t interested in providing a general purpose markdown language server", much annoyance could have been avoided on all sides. Because that is _exactly_ what I’m reading, and where I will finish reading as I’m unsubscribing from this. Further response will not be productive. I seriously wish you the best, but I have exactly zero interest in helping maintain anything in your ecosystem.

**Reply by ChristianMurphy** [reply]
@halostatue your hostility is absolutely uncalled for here, chill dude.
You seem more interested in ranting about straw-man arguments, than solving the problem at hand.
As was previously and repeatedly noted there are several viable options.

> I have exactly zero interest in helping maintain anything in your ecosystem.

This is open source, if you want something to change, you help contribute.

**Reply by remcohaszing** [reply]
To get started, I acknowledge the issue @halostatue is running into. I think he falls in a category of users the language server should support, but currently doesn’t: someone who doesn’t use remark, nor desires to, but who is prompted to install `remark-language-server` by their editor. I did not anticipate this use case when I first started developing `remark-language-server`, unfortunately this is now a negative user experience and the issue has been long standing.

There’s an open issue for this: #30. I believe this is a very important issue to resolve soon and would love to work on it. Beware this is something I’ve worked on in my spare time. Unfortunately last month has been really busy for me, so I haven’t been able to contribute. I expect to have more time to work on this starting next week.

Once everything’s done @halostatue should no longer be bothered by `remark-language-server`, unless they actually integrate remark into their workflow, in which case using remark is a delight. :)

**Reply by chancez** [reply]
> I see several potential answers, several of these already mentioned:
> ....
> * Probably some more? Is this enough?


How about an option to just work with a "built-in" remark (as part of the language-server's dependencies) if no remark is found in a project's node_modules or to use a globally installed remark in $PATH?

**Comment by chancez** [comment]
I am also currently running into this issue over here: https://github.com/williamboman/nvim-lsp-installer/issues/519

I generally just want to install `remark` globally, or not at all ideally, and just have the remark-language-server work in any project, anywhere with markdown, regardless of whether or not `remark` is in a local `node_modules` directory. Having to `npm install remark` in every project I use is an awful user experience and I'd almost rather not use the language server at all since it comes up every single time I work on a new project and edit the README.md

**Comment by axgkl** [comment]
@chancez  @halostatue 

my workaround - ugly but works:

remark will search upwards your path to find its "local installation", ignoring all `node_modules` dirs w/o remark. So go to your home and install it there:  

```bash
cd
npm install remark
```

Also, same style:

```bash
~ ❯ cat .remarkrc.yml  # my example conf, also global:
settings:
  bullet: "-"
  rule: "-"
```

yeah, you end up with a funny `node_modules` folder in your `$HOME` but the formatting I get thru remark made me swallow that pill.

**Reply by halostatue** [reply]
Yeah. Ugly enough that I wouldn’t ever use it, because it breaks all the conventions. After this discussion, I gave up on remark and honestly don’t miss the formatting compared to what I get from `prettier`.

**Reply by axgkl** [reply]
yeah, you were not exactly in sync here. Guess I'll ditch it as well at the next system clean up cycle, when the initial high of having hammered it to sort of working goes away. 
Or symlink it over before editing markdown files, whatever.

Btw, funny: Did not even know that prettier can do markdown, so checked there ... and guess how they do it:

![2023-08-29_1101x272_scrot](https://github.com/remarkjs/remark/assets/9989548/38d2423c-80d4-497d-9036-ecdfd574d925)

;-)

**Reply by wooorm** [reply]
Using `node_modules` folders is not a hack. That’s how the JS ecosystem works.

Not all projects are the same project. Not all projects cater to the same users.
If you want to use prettier, use prettier. If you want to build prettier, or build other things, use remark.
It’s called `remark-language-server`. Not `markdown-language-server` or `prettier-language-server`.
